services/miner/cpuminer.go

```diff
@@ -340,25 +367,85 @@ func (m *CPUMiner) solveBlock(msgBlock *types.Block, ticker *time.Ticker, quit c
 			default:
 				// Non-blocking select to fall through
 			}
-
+			instance := pow.GetInstance(pow.BLAKE2BD,0,[]byte{})
+			powStruct := instance.(*pow.Blake2bd)
 			// Update the nonce and hash the block header.
-			header.Nonce = i
-			h := header.BlockHash()
+			powStruct.Nonce = i
+
+			header.Pow = powStruct
+
 			// Each hash is actually a double hash (tow hashes), so
 			// increment the number of hashes by 2
 			hashesCompleted += 2
+			h := hash.DoubleHashH(header.BlockData())
+			hashNum := pow.HashToBig(&h)
 
+			if hashNum.Cmp(target) > 0 {
+				str := fmt.Sprintf("block hash of %064x is higher than"+
+					" expected max of %064x", hashNum, target)
+				log.Debug(str)
+				continue
+			}
 			// The block is solved when the new block hash is less
 			// than the target difficulty.  Yay!
-			if blockchain.HashToBig(&h).Cmp(targetDifficulty) <= 0 {
-				m.updateHashes <- hashesCompleted
-				return true
-			}
+			m.updateHashes <- hashesCompleted
+			return true
 		}
 	//}
 	return false
 }
 
+// solveBlock attempts to find 42 circles that hash match the target diff
+func (m *CPUMiner) solveCuckarooBlock(msgBlock *types.Block, ticker *time.Ticker, quit chan struct{},scale uint64) bool {
+
+	// Create a couple of convenience variables.
+	header := &msgBlock.Header
+	// Initial state.
+	hashesCompleted := uint64(0)
+	// Search through the entire nonce range for a solution while
+	// periodically checking for early quit and stale block
+	// conditions along with updates to the speed monitor.
+	for i := uint32(0); i <= maxNonce; i++ {
+		select {
+		case <-quit:
+			return false
+
+		case <-ticker.C:
+			m.updateHashes <- hashesCompleted
+			return true
+		default:
+			// Non-blocking select to fall through
+		}
+		instance := pow.GetInstance(pow.CUCKAROO,0,[]byte{})
+		powStruct := instance.(*pow.Cuckaroo)
+		powStruct.Nonce = i
+		// Update the nonce and hash the block header.
+		header.Pow = powStruct
+		powStruct.SetEdgeBits(uint8(cuckoo.Edgebits))
+		sipH := hash.HashH(header.BlockData())
+		c:= cuckoo.NewCuckoo()
+		cycleNonces, isFound := c.PoW(sipH[:])
+		if !isFound {
+			continue
+		}
+		powStruct.SetCircleEdges(cycleNonces)
+		powStruct.SetScale(uint32(scale))
+		err := cuckoo.VerifyCuckaroo(sipH[:],cycleNonces[:],uint(cuckoo.Edgebits))
+		if err != nil{
+			log.Debug(err.Error())
+			continue
+		}
+		if pow.CalcCuckooDiff(int64(scale),powStruct.GetBlockHash([]byte{})) < uint64(header.Difficulty){
+			log.Debug("cuckoo hash difficulty is too easy!")
```

Comment:
same with the above comments, unnecessary debug, show nothing necessary & only slow down the calculation.